### PR TITLE
feat(route53): added L2 construct for Route53's health checks

### PR DESCRIPTION
**Pull request by wladyslawczyzewski**
### Issue # (if applicable)

Part of #9481 
Another PR for this ticket #30664 

### Reason for this change

Added L2 construct for `AWS::Route53::HealthCheck` resource

### Description of changes

The changes only introduces the L2 construct for Route53 health check resources. Except the L2 construct itself, I added basic validations for the input props.

### Description of how you validated changes

- unit tests
- integration tests

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:305]: Switching to a map would also remove the need for this exception, since the `rules` type will require that all possible `HealthCheckType` values are covered

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/README.md:225]: A brief description of what health checks are and what their purpose is would be ideal here, even if failovers aren't supported yet

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/README.md:252]: I would just add a link to the `HealthCheckType` enum doc page. It will remain up to date, and have the corresponding description for each type:

```suggestion
See [`HealthCheckType`](aws-cdk-lib.aws_route53.HealthCheckType.html) for possible values.
```

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:197]: Even though this is not well documented, I'm assuming that `childHealthChecks` is a list of `Ref`s/`healthCheckId`s. I would add an integration test to verify this assumption, and change this type from `string[]` to `IHealthCheck[]`. CDK users will then be able to create multiple health checks in the same stack, or import them with `fromHealthCheckId`.

This should also be able to handle dependency resolution (e.g. if you create both `hc1` and `hc2`, with `hc2` being a child of `hc1`,) but again, that's something you can verify with an integration test

File: packages/@aws-cdk-testing/framework-integ/test/aws-route53/test/integ.health-check.ts
Comment (nmussy) [inline packages/@aws-cdk-testing/framework-integ/test/aws-route53/test/integ.health-check.ts:75]: I would add more properties to your health checks, and make assertions to make sure all the health checks properties are correctly set, see [AwsApiCall](https://docs.aws.amazon.com/cdk/api/v2/docs/integ-tests-alpha-readme.html#api-calls). This way you'll be able to ensure that everything is able to be set properly

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:43]: I strongly suspect the default is `false` here, but this is not made clear in either the Cfn nor the API documentation

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:91]:
```suggestion
   * @default false
```

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:105]: Again, not explicitly said, but I'm going to assume the default is `false`

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:34]: "Complex types" don't make a ton of sense in the CDK, this is just an array

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:119]: The default value is documented in the Cfn docs. Also, "complex type" and "Region element" sound a bit off here. I assume it's just "an array of region identifiers", but you can add this to your integration tests to make sure

```suggestion
   * An array of region identifiers from which you want Amazon Route 53 health checkers to check the specified endpoint.
   *
   * @default - performs checks from all of the regions that are listed under Valid Values.
```

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:126]: "30" is not a valid `Duration`, and we do ask CDK users for a number of seconds here, just a duration

```suggestion
   * The duration between the time that Amazon Route 53 gets a response from your endpoint and the time that it sends the next health check request. Each Route 53 health checker makes requests at this interval.
   *
   * @default Duration.seconds(30)
```

**Review by nmussy — CHANGES_REQUESTED** [review]
Apologies for the number of requested changes, but it's a really good first pass. I would just be mindful of copy-pasting CloudFormation docs for the property JSDocs, some of them don't make as much sense in the context of the CDK 👍

File: packages/aws-cdk-lib/aws-route53/test/record-set.test.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/test/record-set.test.ts:214]: I would also add an assertion for a `AWS::Route53::HealthCheck`

File: packages/aws-cdk-lib/aws-route53/test/record-set.test.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/test/record-set.test.ts:191]: I would also add a test to make sure that imported health checks can be properly mapped to Record sets

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: My main issue is that it's an optional Cfn parameter, and by having a default value here you are preventing CDK users from being able to pass a null-ish value. I'll let the maintainer weigh in here, I don't think it's a huge issue either way

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: See https://github.com/aws/aws-cdk/pull/30739#discussion_r1671724994

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts:617]: Again, I'll leave it up to the maintainer review

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts:514]: You could simplify the tests by doing the following, the first error message also implying a non-empty string value:

```ts
  if (!props.searchString {
    throw new Error(`SearchString is required for health check type: ${props.type}`);
  }

  if (props.searchString.length > 255) {
    throw new Error('SearchString length cannot be greater than 255 characters');
  }
```

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:192]: I have the same maintainability concerns for this JSDoc and this type, I would just revert this type to a `string[]` and add a link to the Cfn documentation.

Just as an FYI, you could have used a `const` assertion to derive this type from the `VALID_REGIONS` array, see this [example Playground](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAagQQDIEkAiB9ASgUQOIoDyAcgMowC8MA2gOQCuEAtAKYCG0TAjLQDQwNmAdxaceAXRgcYoSFADcAKGUBLMFBYAnAGZtgLGAAVNIAA4QYAb0UxbMTSwDmK8BABcMABRQAnqZYg2vDI6Nj4RGTUYPQAtgBGWuIAlNTiSgC+yrLQMABubAA2KgAmHsZmFlSWDs6uHnSMrBxQ3LTi6UrZsGr5RaVGJuaUVjUukPWCTZwATG0dQA)

File: packages/@aws-cdk-testing/framework-integ/test/aws-route53/test/integ.health-check.ts
Comment (nmussy) [reply packages/@aws-cdk-testing/framework-integ/test/aws-route53/test/integ.health-check.ts]: The integ looks great, thanks for being thorough 👍

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: Again, I don't think setting a default value in the CDK is that useful

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: See comment about setting the default value: https://github.com/aws/aws-cdk/pull/30739#discussion_r1671755028

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: See comment about setting the default value: https://github.com/aws/aws-cdk/pull/30739#discussion_r1671755028

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (nmussy) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: Thanks for the doc update, I've added some comments about the typing and validation though:
https://github.com/aws/aws-cdk/pull/30739/files#r1671746172
https://github.com/aws/aws-cdk/pull/30739/files#r1671739388

**Review by nmussy — CHANGES_REQUESTED** [review]
Thanks for the changes, I had a few more comments

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (wladyslawczyzewski) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts]: @nmussy thanks for this, updated the implementation ✅

**Comment by wladyslawczyzewski** [comment]
hi @nmussy , thanks for you review!

I made the regions weak-typed `string[]` property and as result passing all the validation responsibility to CloudFormation itself, however, I think having this prop strong-typed will be better option even if this will require more maintainer work in the future – not all AWS' regions are used for health checks, therefore user may provide the wrong value and will get the error only on deployment phase.

On the default values stuff – you're having a valid points why to not provide default values in the CDK codebase, however, each of that properties have implicit default values, imho, it's better to let users know what the default value is and if they indeed want to have the resource configured differently, they will need to provide their configuration explicitly in the code rather then relying on CloudFormation hidden decisions.

I pushed few changes according to your last review, could you please take a look and let me know if you see anything else that could be improved?

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (nmussy) [inline packages/aws-cdk-lib/aws-route53/README.md:253]: I'm not sure using a zero weight here is the best example, it's a pretty complex behavior with a few pitfalls, see [Weight](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-weighted.html#rrsets-values-weighted-weight) and [How Amazon Route 53 chooses records when health checking is configured](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/health-checks-how-route-53-chooses-records.html)

```suggestion
  weight: 50,
```

**Review by nmussy — APPROVED** [review]
LGTM, I'll let the maintainer make a decision regarding the default values. I'm also not sure using `weight: 0` is a great example, see comment

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (shikha372) [inline packages/aws-cdk-lib/aws-route53/README.md:228]:
```suggestion
In the following example, the `ARecord` will be returned by Route 53 in response to DNS queries only if the HTTP requests to the `example.com/health` endpoint return a 2XX or 3XX status code.
```

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (shikha372) [inline packages/aws-cdk-lib/aws-route53/README.md:226]:
```suggestion
In CDK, this can be done using `HealthCheck` construct.
```

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (shikha372) [inline packages/aws-cdk-lib/aws-route53/README.md:225]: Please link the documentation to route53 health check

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (shikha372) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:292]: instead of having all these functions to validate these scenarios, can we include something like a enum based properties or class to define these props, i think one of the [design principles](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#property-documentation) we follow is to prevent many check cases.

The fewer “if statements” the better.

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (shikha372) [inline packages/aws-cdk-lib/aws-route53/README.md:257]: I would suggest adding these documentation links upward so that its easier to get the overview before looking at the implementation in CDK

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (wladyslawczyzewski) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts:292]: sorry @shikha372 could you please explain this further?

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (wladyslawczyzewski) [reply packages/aws-cdk-lib/aws-route53/README.md]: @shikha372 moved this link to the top of the Health Checks section

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (shikha372) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts:292]: referring to all the validation functions defined with each healthcheck type eg. 
```
function ruleAlarmIdentifierIsNotAllowed(props: HealthCheckProps) {
  if (props.alarmIdentifier) {
    throw new Error(`AlarmIdentifier is not supported for health check type: ${props.type}`);
  }
}
```
instead of validating these checks with if-else, can we introduce a enum type class or method which will allow users to set only the required fields possible under a healthcheck, let me know if you think this is not feasible

File: packages/aws-cdk-lib/aws-route53/README.md
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-route53/README.md:233]:
```suggestion
In the following example, the `ARecord` will be returned by Route 53 in response to DNS queries only if the HTTP requests to the `example.com/health` endpoint return a 2XX or 3XX status code.
```

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:442]: Is the 3 here cloudformation default? i.e. even if you set it to `undefined` for this type of health check, CFN will default it to 3?

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:452]: same as above.

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:320]: I appreciate the fact that you added all these validations on CDK side, however, this also makes me a bit worried at the same time. If CloudFormation decides to change one or some of the validation rules, we need to make changes accordingly and given there are so many, it's not straightforward to update this.

How many of these validations rule are essential? Can we only validate on the essential ones and leave the rest to CFN?

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-route53/lib/health-check.ts:604]: This regex is not human readable IMO. We should leave it out and let CFN to do the validation or please add documentations that link to the CFN page and describe the Regex.

**Review by GavinZZ — CHANGES_REQUESTED** [review]
Hi sorry for keep you waiting on this PR. It's great work and it's a lot of work done and I'll try to help you to get this merged ASAP.

Left some comments to get your thoughts.

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (wladyslawczyzewski) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts:334]: hi @GavinZZ , yes, this is information from the official CFN's docs for this resource type https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-healthcheck-healthcheckconfig.html

File: packages/aws-cdk-lib/aws-route53/lib/health-check.ts
Comment (wladyslawczyzewski) [reply packages/aws-cdk-lib/aws-route53/lib/health-check.ts:344]: hi @GavinZZ , please take a look on my another response to similar comment

**Comment by wladyslawczyzewski** [comment]
hi @GavinZZ , could you please re-review the PR – I've deleted the props validation piece as it was too opinionated, I prefer to have the validation to be in synthesis phase instead deployment as this reduces the feedback time by a lot, but I also see the arguments against this.

**Comment by GavinZZ** [comment]
> hi @GavinZZ , could you please re-review the PR – I've deleted the props validation piece as it was too opinionated, I prefer to have the validation to be in synthesis phase instead deployment as this reduces the feedback time by a lot, but I also see the arguments against this.

IMO I am usually in favour of synthesis validation rather than deployment validation. However, the validation was too complicated and I think we should only include the essential/most obvious ones in the synthesis validation. 

Either way, I think it's a two way door to release this feature without the validation first and we can always add more validation in the future.

**Review by GavinZZ — APPROVED** [review]

**Pull request merged**
